<a href="https://colab.research.google.com/github/allaboutsimple/aibootcamp/blob/main/ds_cs_N421a_ipynb%EC%9D%98_%EC%82%AC%EB%B3%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src='https://user-images.githubusercontent.com/6457691/90080969-0f758d00-dd47-11ea-8191-fa12fd2054a7.png' width = '200' align = 'right'>

## *DATA SCIENCE / SECTION 4 / SPRINT 1 / NOTE 1 - assignmnet*

---





# N411. 인공신경망(Artificial Neural Networks) 과제

## 실제 데이터 과제
 - 아래 주어진 데이터를 신경망을 이용하여 Classification 문제를 풀어보세요.
 - 또한 머신러닝에서 배운 방법(배우지 않은 머신러닝 방법론(SVM 등)도 가능)을 이용하여 비교해보세요.

입력 데이터 샘플과 Features : 1077 샘플 x 69 Features (변수)

데이터 label: 다운증후군 (1), 정상군 (2)

데이터는 다운증후군과 정상군 마우스 피질의 핵 분획에서 검출 가능한 신호를 생성하는 69 개 단백질의 발현 수준으로 구성되어 있습니다.
라벨로는 다운증후군 1, 정상군 2로 할당되어 있습니다.


In [1]:
!pip install tensorflow-gpu==2.0.0-rc1
import tensorflow as tf

     |████████████████████████████████| 380.5 MB 9.0 kB/s 
     |████████████████████████████████| 50 kB 7.1 MB/s 
     |████████████████████████████████| 501 kB 77.2 MB/s 
     |████████████████████████████████| 4.3 MB 35.0 MB/s 


In [29]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf
from sklearn.model_selection import train_test_split

df = pd.read_excel("https://ds-lecture-data.s3.ap-northeast-2.amazonaws.com/MouseProtein/mouse_protein_X.xls", header=None)
df_label = pd.read_excel("https://ds-lecture-data.s3.ap-northeast-2.amazonaws.com/MouseProtein/mouse_protein_label.xls", header=None)

In [30]:
# 샘플당 100개의 특성(feature)을 가진 데이터
df.head()

0        1        2       3   ...       65       66       67      68
0  0.50364  0.74719  0.43018  2.8163  ...  1.04500  0.83156  0.18885  1.6757
1  0.51462  0.68906  0.41177  2.7895  ...  1.00990  0.84927  0.20040  1.7436
2  0.50918  0.73025  0.41831  2.6872  ...  0.99685  0.84671  0.19368  1.9264
3  0.44211  0.61708  0.35863  2.4669  ...  0.99022  0.83328  0.19211  1.7006
4  0.43494  0.61743  0.35880  2.3658  ...  0.99777  0.87867  0.20560  1.8397

[5 rows x 69 columns]

In [31]:
print(df_label.head())
print(df_label.tail())

   0
0  1
1  1
2  1
3  1
4  1
      0
1072  2
1073  2
1074  2
1075  2
1076  2


In [32]:
# 타겟 레이블을 정상군 = 0, 다운증후군 = 1로 변경 
y = np.where(df_label == 2, 0, 1)
X = df
y

array([[1],
       [1],
       [1],
       ...,
       [0],
       [0],
       [0]])

In [33]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size =0.2, random_state = 42)

In [36]:
# 신경망 모델 구축

model = tf.keras.models.Sequential([
                                    tf.keras.layers.Dense(65, activation='relu', input_dim = 69),
                                    tf.keras.layers.Dense(65, activation='relu'),
                                    tf.keras.layers.Dense(1, activation= 'sigmoid')
                                    ])

model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])
# train셋으로 학습
model.fit(X_train, y_train, epochs=60,validation_split=0.2, verbose=2)

# 모델을 이용하여 예측
model.evaluate(x_test,  y_test, verbose=2)

Train on 688 samples, validate on 173 samples
Epoch 1/60
688/688 - 0s - loss: 0.6807 - accuracy: 0.5843 - val_loss: 0.6652 - val_accuracy: 0.6474
Epoch 2/60
688/688 - 0s - loss: 0.6510 - accuracy: 0.6265 - val_loss: 0.6358 - val_accuracy: 0.6705
Epoch 3/60
688/688 - 0s - loss: 0.6315 - accuracy: 0.6294 - val_loss: 0.6160 - val_accuracy: 0.6532
Epoch 4/60
688/688 - 0s - loss: 0.6119 - accuracy: 0.6860 - val_loss: 0.5918 - val_accuracy: 0.6994
Epoch 5/60
688/688 - 0s - loss: 0.5829 - accuracy: 0.7035 - val_loss: 0.5861 - val_accuracy: 0.6994
Epoch 6/60
688/688 - 0s - loss: 0.5721 - accuracy: 0.6977 - val_loss: 0.5496 - val_accuracy: 0.7052
Epoch 7/60
688/688 - 0s - loss: 0.5419 - accuracy: 0.7442 - val_loss: 0.5331 - val_accuracy: 0.7110
Epoch 8/60
688/688 - 0s - loss: 0.5090 - accuracy: 0.7762 - val_loss: 0.5081 - val_accuracy: 0.7168
Epoch 9/60
688/688 - 0s - loss: 0.4844 - accuracy: 0.7863 - val_loss: 0.4858 - val_accuracy: 0.7283
Epoch 10/60
688/688 - 0s - loss: 0.4779 - accuracy: 0.

[0.20910107758310106, 0.9537037]

In [48]:
from sklearn.metrics import classification_report as clrpt
y_pred = model.predict(X_test)
y_pred = np.round(y_pred)
print(clrpt(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.90      0.95        97
           1       0.92      1.00      0.96       119

    accuracy                           0.95       216
   macro avg       0.96      0.95      0.95       216
weighted avg       0.96      0.95      0.95       216



In [49]:
# 답 제출 1

from sklearn.metrics import *

accuracy = accuracy_score(y_test, y_pred),
precision = precision_score(y_test, y_pred),
recall = recall_score(y_test, y_pred),
f1 = f1_score(y_test, y_pred)

print(f'Accuracy: {accuracy[0]:.4f},\nPrecision: {precision[0]:.4f},\nRecall: {recall[0]:.4f},\nF1_score: {f1:.4f}')

Accuracy: 0.9537,
Precision: 0.9225,
Recall: 1.0000,
F1_score: 0.9597


In [53]:
#머신러닝
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import make_pipeline
from sklearn.impute import SimpleImputer

# 파이프라인 생성
pipe = make_pipeline(
    SimpleImputer(),
    RandomForestClassifier(n_estimators = 100, random_state =2, n_jobs = -1)
)

# 데이터 세트 분리
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

# fit
pipe.fit(X_train, y_train)
y_pred = pipe.predict(X_test)

# 점수확인
print(clrpt(y_test, y_pred))

/usr/local/lib/python3.7/dist-packages/sklearn/pipeline.py:354: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  self._final_estimator.fit(Xt, y, **fit_params)


              precision    recall  f1-score   support

           0       1.00      0.96      0.98        97
           1       0.97      1.00      0.98       119

    accuracy                           0.98       216
   macro avg       0.98      0.98      0.98       216
weighted avg       0.98      0.98      0.98       216



In [52]:
# 답 제출 2

from sklearn.metrics import *

accuracy = accuracy_score(y_test, y_pred),
precision = precision_score(y_test, y_pred),
recall = recall_score(y_test, y_pred),
f1 = f1_score(y_test, y_pred)

print(f'Accuracy: {accuracy[0]:.4f},\nPrecision: {precision[0]:.4f},\nRecall: {recall[0]:.4f},\nF1_score: {f1:.4f}')

Accuracy: 0.9815,
Precision: 0.9675,
Recall: 1.0000,
F1_score: 0.9835


---

4-1. 사용한 모델을 입력합니다. 

4-2. Accuracy를 입력합니다. 

4-3. Precision 을 입력합니다. 

4-4. Recall 을 입력합니다.

4-5. F1 score 를 입력합니다. 

(신경망, 0.500, 0.368, 0.632, 4.299) 와 같이 입력합니다.